drv/sprot-api/src/lib.rs

```diff
@@ -3,64 +3,266 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
 //! Client API for SP to RoT messages over SPI.
-//!
-//! An SP/RoT SPI message is:
-//!   1. A hubpack encoded `MsgHeader` containing the protocol, payload length,
-//!      and message type.
-//!   2. Payload according to MessageType, typically hubpack encoded
-//!      structure(s) and/or bulk data.
-//!   3. A CRC16 value that covers all of the bytes from the protocol ID to
-//!      the end of the payload.
-//!
 
 #![no_std]
 extern crate memoffset;
 
 use crc::{Crc, CRC_16_XMODEM};
-use derive_idol_err::IdolError;
+use derive_more::From;
 use drv_spi_api::SpiError;
 pub use drv_update_api::{
     HandoffDataLoadError, RotBootState, RotSlot, SlotId, SwitchDuration,
     UpdateError, UpdateTarget,
 };
+use dumper_api::DumperError;
 use hubpack::SerializedSize;
-use idol_runtime::{Leased, R};
+use idol_runtime::{Leased, LenLimit, RequestError, R};
 use serde::{Deserialize, Serialize};
-use userlib::{sys_send, FromPrimitive};
-use zerocopy::{AsBytes, FromBytes};
-#[cfg(feature = "sink_test")]
-use zerocopy::{ByteOrder, LittleEndian};
-
-/// The canonical SpRot protocol error returned by this API
-//
-// TODO: Audit that each MsgError is used and has some reasonable action.
-// While a diverse set of error codes may be useful for debugging it
-// clutters code that just has to deal with the error.
-// then consider adding a function that translates an error code
-// into the desired action, e.g. InvalidCrc and FlowError should both
-// result in a retry on the SP side and an ErrorRsp on the RoT side.
-// Note: Variant order must be maintained to stay compatible due to SP/RoT
-// version skew during update.
+use sprockets_common::msgs::{
+    RotError as SprocketsError, RotRequestV1 as SprocketsReq,
+    RotResponseV1 as SprocketsRsp,
+};
+use static_assertions::const_assert;
+use userlib::sys_send;
+
+const CRC16: Crc<u16> = Crc::<u16>::new(&CRC_16_XMODEM);
+pub const CRC_SIZE: usize = <u16 as SerializedSize>::MAX_SIZE;
+// XXX ROT FIFO size should be discovered.
+pub const ROT_FIFO_SIZE: usize = 8;
+pub const MAX_BLOB_SIZE: usize = 512;
+pub const MAX_REQUEST_SIZE: usize =
+    Header::MAX_SIZE + ReqBody::MAX_SIZE + MAX_BLOB_SIZE;
+pub const MAX_RESPONSE_SIZE: usize =
+    Header::MAX_SIZE + RspBody::MAX_SIZE + MAX_BLOB_SIZE;
+
+// For simplicity we want to be able to retrieve the header
+// in a maximum of 1 FIFO size read.
+const_assert!(Header::MAX_SIZE <= ROT_FIFO_SIZE);
+
+pub type Request = Msg<ReqBody>;
+pub type Response = Msg<Result<RspBody, SprotError>>;
+
+/// A message header for a request or response
+///
+/// It's important that this header be kept fixed size by limiting the use
```

Comment:
What does this comment mean?

I think it means that the header needs to contain all fields with fixed size serialization in Hubpack, which basically means primitives, structs, or simple enums?

Reply:
Precisely. I will clarify.